keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,565 @@
+  cannot	be used for work queue patterns.	
+- The `.spec.parallelism` must be up to `10^6`. This is to guarantee that we
+  can keep track of completions per-index in the Job status in the future.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Indexed Jobs get the annotation `batch.alpha.kubernetes.io/job-completion-index`
+with a value equal to its completion index. The annotation is immutable.
+
+The annotation can be accessed through the downward API as a file or environment
+variable.
+
+For user convenience, the Job controller adds the completion index as an
+environment variable through the downward API. That is, the Job controller
+creates Pods like so:
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+    - name: test-container
+      env:
+        - name: JOB_COMPLETION_INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['batch.alpha.kubernetes.io/job-completion-index'] 
+```
+
+The Job controller doesn't add the environment variable if there is a name
+conflict with an existing environment variables in the Job's Pod template.
+
+### Job completion and restart policy
+
+When dealing with Indexed Jobs, the Job controller keeps track of Pod
+completions for each index from 0 to `.spec.completions - 1`.
+Once the controller notices that a Pod finishes successfully, it will not create
+another Pod with the same index.
+
+The Job controller considers a Job completed when there is at least one
+successful Pod for each completion index.
+
+If an entire Pod fails, such as when the Pod gets kicked off the node or if a
+container of the Pod fails and `.spec.template.spec.restartPolicy = "Never"`,
+the Job controller gets the completion index for the failed Pod and creates a
+new Pod with the same index. The application needs to handle restarts in a
+different Pod.
+
+The Pod might not be immediately replaced, as the number of active Pods could
+have hit the parallelism limit. Once another Pod finishes, the Job controller
+syncs the Job, scanning for unsatisfied indexes and creates the missing Pods for
+them.
+
+The kubelet handles container restarts as usual, according to the
+`spec.template.spec.restartPolicy`.
+
+<<[UNRESOLVED TBD Beta: Track completed indexes in Job status]>>
+Once [kubernetes/kubernetes#28486](https://github.com/kubernetes/kubernetes/issues/28486)
+is resolved:
+
+The Job controller keeps track of completed indexes in the Job status in a
+base64 encoded bitmask.
+The command `kubectl describe` displays the list of completed indexes in a
+human-readable compressed format. For example:
+
+```
+Completed Indexes: [1-25,28,30-32]
+``` 
+
+The command crops the list of indexes once it reaches 200 characters.
+
+```
+Completed Indexes: [1-25,28,30-32,...]
+``` 
+<<[/UNRESOLVED]>>
+
+### Job scaling
+
+Jobs can be scaled up and down by varying `.spec.parallelism` (note that
+`.spec.completions` is an immutable field).
```

Comment:
Please avoid using "scale" here, as Jobs don't follow the normal semantics for scaling. Ref https://github.com/kubernetes/kubernetes/pull/60139

Reply:
Fixed